node/src/store/utils.rs

```diff
@@ -13,34 +13,48 @@ pub(crate) const VALIDATIONS_PER_YIELD: usize = 4;
 pub(crate) fn calculate_range_to_fetch(
     head_height: u64,
     store_headers: &[RangeInclusive<u64>],
+    syncing_window_edge: Option<u64>,
     limit: u64,
+) -> HeaderRange {
+    let mut missing_range = get_most_recent_missing_range(head_height, store_headers);
+
+    // truncate to syncing window, if height is known
+    if let Some(window_edge) = syncing_window_edge {
+        if missing_range.start() < &window_edge {
+            missing_range = window_edge + 1..=*missing_range.end();
+        }
+    }
+
+    // truncate number of headers to limit
+    if missing_range.len() > limit {
+        let end = missing_range.end();
+        let start = end.saturating_sub(limit) + 1;
+        missing_range = start..=*end;
+    }
+
+    missing_range
+}
+
+fn get_most_recent_missing_range(
+    head_height: u64,
+    store_headers: &[RangeInclusive<u64>],
 ) -> HeaderRange {
     let mut store_headers_iter = store_headers.iter().rev();
 
     let Some(store_head_range) = store_headers_iter.next() else {
-        // empty store, just fetch from head
-        return head_height.saturating_sub(limit) + 1..=head_height;
+        // empty store, we're missing everything
+        return 1..=head_height;
     };
 
     if store_head_range.end() != &head_height {
```

Comment:
This shouldn't be `store_head_range.end() < &head_height`?

Reply:
Store head height > network head height, so probably something already went bit wrong, but I don't suppose it'd hurt here.